// appium-android-driver/build/lib/commands/general.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseWindows = parseWindows;
exports.parseSurfaceLine = parseSurfaceLine;
exports.default = exports.helpers = exports.commands = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _lodash = _interopRequireDefault(require("lodash"));

var _androidHelpers = _interopRequireDefault(require("../android-helpers"));

var _appiumSupport = require("appium-support");

var _bluebird = _interopRequireDefault(require("bluebird"));

var _logger = _interopRequireDefault(require("../logger"));

var _webviewHelpers = require("../webview-helpers");

var _moment = _interopRequireDefault(require("moment"));

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

commands.keys = function () {
  var _ref = (0, _asyncToGenerator2.default)(function* (keys) {
    keys = _lodash.default.isArray(keys) ? keys.join('') : keys;
    let params = {
      text: keys,
      replace: false
    };

    if (this.opts.unicodeKeyboard) {
      params.unicodeKeyboard = true;
    }

    yield this.doSendKeys(params);
  });

  return function (_x) {
    return _ref.apply(this, arguments);
  };
}();

commands.doSendKeys = function () {
  var _ref2 = (0, _asyncToGenerator2.default)(function* (params) {
    return yield this.bootstrap.sendAction('setText', params);
  });

  return function (_x2) {
    return _ref2.apply(this, arguments);
  };
}();

const DEFAULT_DEVICE_TIME = 'YYYY-MM-DDTHH:mm:ssZ';

commands.getDeviceTime = function () {
  var _ref3 = (0, _asyncToGenerator2.default)(function* (format = DEFAULT_DEVICE_TIME) {
    _logger.default.info('Attempting to capture android device date and time');

    if (!_lodash.default.isString(format)) {
      _logger.default.errorAndThrow(`The format specifier is expected to be a valid string specifier like '${DEFAULT_DEVICE_TIME}'. ` + `The current value is: ${format}`);
    }

    const deviceTimestamp = (yield this.adb.shell(['date', '+%Y-%m-%dT%T%z'])).trim();
    const parsedTimestamp = (0, _moment.default)(deviceTimestamp, 'YYYY-MM-DDTHH:mm:ssZ');

    if (!parsedTimestamp.isValid()) {
      _logger.default.warn(`Cannot parse the returned timestamp '${deviceTimestamp}'. Returning as is`);

      return deviceTimestamp;
    }

    return parsedTimestamp.format(format);
  });

  return function () {
    return _ref3.apply(this, arguments);
  };
}();

commands.getPageSource = (0, _asyncToGenerator2.default)(function* () {
  return yield this.bootstrap.sendAction('source');
});
commands.back = (0, _asyncToGenerator2.default)(function* () {
  return yield this.bootstrap.sendAction('pressBack');
});
commands.isKeyboardShown = (0, _asyncToGenerator2.default)(function* () {
  let keyboardInfo = yield this.adb.isSoftKeyboardPresent();
  return keyboardInfo.isKeyboardShown;
});
commands.hideKeyboard = (0, _asyncToGenerator2.default)(function* () {
  let _ref8 = yield this.adb.isSoftKeyboardPresent(),
      isKeyboardShown = _ref8.isKeyboardShown,
      canCloseKeyboard = _ref8.canCloseKeyboard;

  if (!isKeyboardShown) {
    throw new Error("Soft keyboard not present, cannot hide keyboard");
  }

  if (canCloseKeyboard) {
    return yield this.back();
  }

  _logger.default.info("Keyboard has no UI; no closing necessary");
});

commands.openSettingsActivity = function () {
  var _ref9 = (0, _asyncToGenerator2.default)(function* (setting) {
    let _ref10 = yield this.adb.getFocusedPackageAndActivity(),
        appPackage = _ref10.appPackage,
        appActivity = _ref10.appActivity;

    yield this.adb.shell(['am', 'start', '-a', `android.settings.${setting}`]);
    yield this.adb.waitForNotActivity(appPackage, appActivity, 5000);
  });

  return function (_x3) {
    return _ref9.apply(this, arguments);
  };
}();

commands.getWindowSize = (0, _asyncToGenerator2.default)(function* () {
  return yield this.bootstrap.sendAction('getDeviceSize');
});
commands.getWindowRect = (0, _asyncToGenerator2.default)(function* () {
  const _ref13 = yield this.getWindowSize(),
        width = _ref13.width,
        height = _ref13.height;

  return {
    width,
    height,
    x: 0,
    y: 0
  };
});
commands.getCurrentActivity = (0, _asyncToGenerator2.default)(function* () {
  return (yield this.adb.getFocusedPackageAndActivity()).appActivity;
});
commands.getCurrentPackage = (0, _asyncToGenerator2.default)(function* () {
  return (yield this.adb.getFocusedPackageAndActivity()).appPackage;
});

commands.background = function () {
  var _ref16 = (0, _asyncToGenerator2.default)(function* (seconds) {
    if (seconds < 0) {
      yield this.adb.goToHome();
      return true;
    }

    let _ref17 = yield this.adb.getFocusedPackageAndActivity(),
        appPackage = _ref17.appPackage,
        appActivity = _ref17.appActivity;

    yield this.adb.goToHome();
    yield _bluebird.default.delay(seconds * 1000);
    let args;

    if (this._cachedActivityArgs && this._cachedActivityArgs[`${appPackage}/${appActivity}`]) {
      args = this._cachedActivityArgs[`${appPackage}/${appActivity}`];
    } else {
      try {
        _logger.default.debug(`Activating app '${appPackage}' in order to restore it`);

        yield this.activateApp(appPackage);
        return true;
      } catch (ign) {}

      args = appPackage === this.opts.appPackage && appActivity === this.opts.appActivity || appPackage === this.opts.appWaitPackage && (this.opts.appWaitActivity || '').split(',').includes(appActivity) ? {
        pkg: this.opts.appPackage,
        activity: this.opts.appActivity,
        action: this.opts.intentAction,
        category: this.opts.intentCategory,
        flags: this.opts.intentFlags,
        waitPkg: this.opts.appWaitPackage,
        waitActivity: this.opts.appWaitActivity,
        optionalIntentArguments: this.opts.optionalIntentArguments,
        stopApp: false,
        user: this.opts.userProfile
      } : {
        pkg: appPackage,
        activity: appActivity,
        waitPkg: appPackage,
        waitActivity: appActivity,
        stopApp: false
      };
    }

    args = yield _appiumSupport.util.filterObject(args);

    _logger.default.debug(`Bringing application back to foreground with arguments: ${JSON.stringify(args)}`);

    return yield this.adb.startApp(args);
  });

  return function (_x4) {
    return _ref16.apply(this, arguments);
  };
}();

commands.getStrings = function () {
  var _ref18 = (0, _asyncToGenerator2.default)(function* (language) {
    if (!language) {
      language = yield this.adb.getDeviceLanguage();

      _logger.default.info(`No language specified, returning strings for: ${language}`);
    }

    if (this.apkStrings[language]) {
      return this.apkStrings[language];
    }

    this.apkStrings[language] = yield _androidHelpers.default.pushStrings(language, this.adb, this.opts);
    yield this.bootstrap.sendAction('updateStrings');
    return this.apkStrings[language];
  });

  return function (_x5) {
    return _ref18.apply(this, arguments);
  };
}();

commands.launchApp = (0, _asyncToGenerator2.default)(function* () {
  yield this.initAUT();
  yield this.startAUT();
});

commands.startActivity = function () {
  var _ref20 = (0, _asyncToGenerator2.default)(function* (appPackage, appActivity, appWaitPackage, appWaitActivity, intentAction, intentCategory, intentFlags, optionalIntentArguments, dontStopAppOnReset) {
    _logger.default.debug(`Starting package '${appPackage}' and activity '${appActivity}'`);

    if (!_appiumSupport.util.hasValue(dontStopAppOnReset)) {
      dontStopAppOnReset = !!this.opts.dontStopAppOnReset;
    }

    let args = {
      pkg: appPackage,
      activity: appActivity,
      waitPkg: appWaitPackage || appPackage,
      waitActivity: appWaitActivity || appActivity,
      action: intentAction,
      category: intentCategory,
      flags: intentFlags,
      optionalIntentArguments,
      stopApp: !dontStopAppOnReset
    };
    this._cachedActivityArgs = this._cachedActivityArgs || {};
    this._cachedActivityArgs[`${args.waitPkg}/${args.waitActivity}`] = args;
    yield this.adb.startApp(args);
  });

  return function (_x6, _x7, _x8, _x9, _x10, _x11, _x12, _x13, _x14) {
    return _ref20.apply(this, arguments);
  };
}();

commands.reset = (0, _asyncToGenerator2.default)(function* () {
  yield _androidHelpers.default.resetApp(this.adb, Object.assign({}, this.opts, {
    fastReset: true
  }));
  this.curContext = _webviewHelpers.NATIVE_WIN;
  return (yield this.isChromeSession) ? this.startChromeSession() : this.startAUT();
});
commands.startAUT = (0, _asyncToGenerator2.default)(function* () {
  yield this.adb.startApp({
    pkg: this.opts.appPackage,
    activity: this.opts.appActivity,
    action: this.opts.intentAction,
    category: this.opts.intentCategory,
    flags: this.opts.intentFlags,
    waitPkg: this.opts.appWaitPackage,
    waitActivity: this.opts.appWaitActivity,
    waitDuration: this.opts.appWaitDuration,
    optionalIntentArguments: this.opts.optionalIntentArguments,
    stopApp: !this.opts.dontStopAppOnReset,
    user: this.opts.userProfile
  });
});

commands.setUrl = function () {
  var _ref23 = (0, _asyncToGenerator2.default)(function* (uri) {
    yield this.adb.startUri(uri, this.opts.appPackage);
  });

  return function (_x15) {
    return _ref23.apply(this, arguments);
  };
}();

commands.closeApp = (0, _asyncToGenerator2.default)(function* () {
  yield this.adb.forceStop(this.opts.appPackage);
  this.curContext = _webviewHelpers.NATIVE_WIN;
  yield this.stopChromedriverProxies();
});
commands.getDisplayDensity = (0, _asyncToGenerator2.default)(function* () {
  let out = yield this.adb.shell(['getprop', 'ro.sf.lcd_density']);

  if (out) {
    let val = parseInt(out, 10);

    if (!isNaN(val)) {
      return val;
    }

    _logger.default.debug(`Parsed density value was NaN: "${out}"`);
  }

  out = yield this.adb.shell(['getprop', 'qemu.sf.lcd_density']);

  if (out) {
    let val = parseInt(out, 10);

    if (!isNaN(val)) {
      return val;
    }

    _logger.default.debug(`Parsed density value was NaN: "${out}"`);
  }

  _logger.default.errorAndThrow('Failed to get display density property.');
});

function parseSurfaceLine(line) {
  let bounds = line.split('rect=')[1].replace(/[(), x]+/g, ' ').trim().split(' ');
  return {
    visible: line.indexOf('shown=true') !== -1,
    x: parseFloat(bounds[0]),
    y: parseFloat(bounds[1]),
    width: parseFloat(bounds[2]),
    height: parseFloat(bounds[3])
  };
}

function parseWindows(lines) {
  let atStatusBar = false;
  let atNavBar = false;
  let statusBar;
  let navigationBar;
  lines.split('\n').forEach(line => {
    if (line.indexOf('  Window #') !== -1) {
      atStatusBar = line.indexOf('StatusBar') !== -1;
      atNavBar = line.indexOf('NavigationBar') !== -1;
      return;
    }

    if (line.indexOf('      Surface:') === -1) {
      return;
    }

    if (atStatusBar) {
      statusBar = parseSurfaceLine(line);
      atStatusBar = false;
    } else if (atNavBar) {
      navigationBar = parseSurfaceLine(line);
      atNavBar = false;
    }
  });

  if (!statusBar) {
    _logger.default.errorAndThrow('Failed to parse status bar information.');
  }

  if (!navigationBar) {
    _logger.default.errorAndThrow('Failed to parse navigation bar information.');
  }

  return {
    statusBar,
    navigationBar
  };
}

commands.getSystemBars = (0, _asyncToGenerator2.default)(function* () {
  let out = yield this.adb.shell(['dumpsys', 'window', 'windows']);

  if (!out) {
    _logger.default.errorAndThrow('Did not get window manager output.');
  }

  return parseWindows(out);
});

commands.mobilePerformEditorAction = function () {
  var _ref27 = (0, _asyncToGenerator2.default)(function* (opts = {}) {
    const action = opts.action;

    if (!_appiumSupport.util.hasValue(action)) {
      _logger.default.errorAndThrow(`'action' argument is required`);
    }

    yield this.adb.performEditorAction(action);
  });

  return function () {
    return _ref27.apply(this, arguments);
  };
}();

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
