// appium-chromedriver/build/lib/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getChromeVersion = getChromeVersion;
exports.getChromedriverDir = getChromedriverDir;
exports.getChromedriverBinaryPath = getChromedriverBinaryPath;
exports.getCurPlatform = getCurPlatform;
exports.getPlatforms = getPlatforms;
exports.CD_BASE_DIR = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _appiumSupport = require("appium-support");

var _path = _interopRequireDefault(require("path"));

var _chromedriver = require("./chromedriver");

const CD_BASE_DIR = _path.default.resolve(__dirname, '..', '..', 'chromedriver');

exports.CD_BASE_DIR = CD_BASE_DIR;

function getChromeVersion(_x, _x2) {
  return _getChromeVersion.apply(this, arguments);
}

function _getChromeVersion() {
  _getChromeVersion = (0, _asyncToGenerator2.default)(function* (adb, bundleId) {
    const _ref = yield adb.getPackageInfo(bundleId),
          versionName = _ref.versionName;

    return versionName;
  });
  return _getChromeVersion.apply(this, arguments);
}

function getChromedriverDir(platform = getCurPlatform()) {
  return _path.default.resolve(CD_BASE_DIR, platform);
}

function getChromedriverBinaryPath() {
  return _getChromedriverBinaryPath.apply(this, arguments);
}

function _getChromedriverBinaryPath() {
  _getChromedriverBinaryPath = (0, _asyncToGenerator2.default)(function* (platform = getCurPlatform(), arch = null) {
    const baseDir = getChromedriverDir(platform);
    let ext = '';

    if (platform === 'win') {
      ext = '.exe';
    } else if (platform === 'linux') {
      ext = `_${arch || (yield _appiumSupport.system.arch())}`;
    }

    return _path.default.resolve(baseDir, `chromedriver${ext}`);
  });
  return _getChromedriverBinaryPath.apply(this, arguments);
}

function getCurPlatform() {
  return _appiumSupport.system.isWindows() ? 'win' : _appiumSupport.system.isMac() ? 'mac' : 'linux';
}

function getPlatforms() {
  let plats = [['win', '32'], ['linux', '64']];
  const cdVer = parseFloat(_chromedriver.CD_VER);
  plats.push(cdVer < 2.23 ? ['mac', '32'] : ['mac', '64']);

  if (cdVer < 2.34) {
    plats.push(['linux', '32']);
  }

  return plats;
}require('source-map-support').install();
